File: IQ/Classes/clsAvalancheOPG.vb
Imports dataAccess

Public Class ClsAvalancheOPG

    'systems carry a list of avalancheOPG's
    Property ID As Integer
    Property OPGref
    Property Region As clsRegion
    Property ValidFrom As DateTime
    Property ValidTo As DateTime
    Property OptMin As Integer
    Property OptMax As Integer
    'Property LPDiscountPercent As Single 'percent discount off list price (prices(HP))
    Property Options As Dictionary(Of Integer, ClsAvalancheOption)

    ReadOnly Property DisplayName(language As clsLanguage) As String
        Get
            DisplayName = OPGref
        End Get
    End Property
    Public Function Insert() As ClsAvalancheOPG

        Dim av As ClsAvalancheOPG = New ClsAvalancheOPG(Me.OPGref, Me.region, Me.ValidFrom, Me.ValidTo, Me.OptMin, Me.OptMax)
        Return av

    End Function

    Public Function getAvalancheOptions(Optional prodref As String = "", Optional qty As Integer = 0, Optional dateTime As Object = Nothing, Optional region As clsRegion = Nothing) As List(Of clsAvalancheOption)

        Pmark("getAvalancheOptions")
        'returns the avalancheOptions (containing % rebate information)  is for the sepcified prodref,qty..etc (which are all optional)

        getAvalancheOptions = New List(Of clsAvalancheOption)

        Dim dateValid As Boolean = False
        If dateTime Is Nothing Then
            dateValid = True
        Else
            If dateTime > Me.ValidFrom And dateTime < Me.ValidTo Then
                dateValid = True
            End If
        End If

        Dim regionValid As Boolean = False
        If region Is Nothing Then
            regionValid = True
        Else
            regionValid = Me.Region.Encompasses(region)
        End If

        If dateValid Then
            If (qty >= Me.OptMin And qty <= Me.OptMax) Or qty = 0 Then
                If regionValid Then
                    For Each o In Me.Options.Values
                        If o.ProdRef = prodref Or prodref = "" Then
                            'avalanche gives a discount as a percentage of list price
                            getAvalancheOptions.Add(o)
                        End If
                    Next
                End If
            End If
        End If

        Pacc("getAvalancheOptions")

    End Function

    Public Sub Update()

        Dim sql$
        sql$ = "UPDATE Avalance SET "
        sql$ &= "opgRef=" & Me.OPGref & ","
        '  sql$ &= "prodRef=" & Me.ProdRef & ","
        sql$ &= "FK_region_id=" & Me.region.ID & ","
        sql$ &= "validFrom=" & da.universaldate(Me.ValidFrom) & ","
        sql$ &= "validTo=" & da.universaldate(Me.ValidTo) & ","
        ' sql$ &= "LPDiscountPercent =" & Me.LPDiscountPercent & ","
        sql$ &= "optMin=" & Me.OptMin & ","
        sql$ &= "optMax=" & Me.OptMax & ""
        sql$ &= " WHERE ID = " & Me.ID

        da.DBExecutesql(sql)

        'TODO : - need to update iq.i_OPGref 

    End Sub

    Public Sub New(opgRef As String, Region As clsRegion, validFrom As DateTime, ValidTo As DateTime, optMin As Integer, optMax As Integer, Optional writecache As DataTable = Nothing)

        Me.OPGref = opgRef
        'Me.ProdRef = prodRef
        Me.region = Region
        Me.ValidFrom = validFrom
        Me.ValidTo = ValidTo
        Me.OptMin = optMin
        Me.OptMax = optMax
        Me.Options = New Dictionary(Of Integer, ClsAvalancheOption)

        If writecache Is Nothing Then

            Dim sql$
            sql$ = "INSERT INTO avalancheOPG (opgref,optmin,optmax,validFrom,validTo,fk_region_id) "
            sql$ &= "VALUES (" & opgRef & "," & optMin & "," & optMax & "," & da.universaldate(validFrom) & "," & da.universaldate(ValidTo) & "," & Region.ID & ")"

            Me.ID = da.DBExecutesql(sql$, True)
            iq.AvalancheOPGs.Add(Me.ID, Me)

        Else

            Dim row As System.Data.DataRow
            row = writecache.NewRow()

            row("opgref") = Me.OPGref
            'row("prodref") = Me.ProdRef
            row("optmin") = Me.OptMin
            row("optmax") = Me.OptMax
            'row("LPDiscountPercent") = Me.LPDiscountPercent
            row("validFrom") = Me.ValidFrom
            row("validTo") = Me.ValidTo
            row("fk_region_id") = Me.region.ID

            writecache.Rows.Add(row)

        End If

        iq.i_OpgRef.Add(opgRef, Me)

        'If a system has a descendant price with  offer which points to a pool contaiining the system

    End Sub
    Public Sub New()

        Me.ID = -1


    End Sub


    Public Sub New(ByVal ID As Integer, opgRef As String, Region As clsRegion, validFrom As DateTime, ValidTo As DateTime, optMin As Integer, optMax As Integer)

        'the OPG's don't carry a set of systems - becuase the systems have the OPG's added (product.avalancheOPGs)... which is how they're needed 
        '(we typically want to know the OPG's for a system - not the systems for an OPG)

        Me.ID = ID
        Me.OPGref = opgRef
        Me.region = Region
        Me.ValidFrom = validFrom
        Me.ValidTo = ValidTo
        Me.OptMin = optMin
        Me.OptMax = optMax

        iq.AvalancheOPGs.Add(Me.ID, Me)
        iq.i_OpgRef.Add(opgRef, Me)

        Me.Options = New Dictionary(Of Integer, ClsAvalancheOption)

    End Sub


End Class

File: IQ/Classes/clsAvalancheOPG.cs
using dataAccess;


public class ClsAvalancheOPG
{

    //systems carry a list of avalancheOPG's
    public int ID { get; set; }
    public dynamic OPGref { get; set; }
    public clsRegion Region { get; set; }
    public DateTime ValidFrom { get; set; }
    public DateTime ValidTo { get; set; }
    public int OptMin { get; set; }
    public int OptMax { get; set; }
    //Property LPDiscountPercent As Single 'percent discount off list price (prices(HP))
    public Dictionary<int, ClsAvalancheOption> Options { get; set; }

    public string get_DisplayName(clsLanguage language)
    {
        string returnValue = "";
        returnValue = System.Convert.ToString(OPGref);
        return returnValue;
    }
    public ClsAvalancheOPG Insert()
    {

        ClsAvalancheOPG av = new ClsAvalancheOPG(System.Convert.ToString(this.OPGref), this.Region, this.ValidFrom, this.ValidTo, this.OptMin, this.OptMax);
        return av;

    }

    public List<clsAvalancheOption> getAvalancheOptions(string prodref = "", int qty = 0, object dateTime = null, clsRegion region = null)
    {
        List<clsAvalancheOption> returnValue = default(List<clsAvalancheOption>);

        Pmark("getAvalancheOptions");
        //returns the avalancheOptions (containing % rebate information)  is for the sepcified prodref,qty..etc (which are all optional)

        returnValue = new List<clsAvalancheOption>();

        bool dateValid = false;
        if (dateTime == null)
        {
            dateValid = true;
        }
        else
        {
            if (dateTime > this.ValidFrom && dateTime < this.ValidTo)
            {
                dateValid = true;
            }
        }

        bool regionValid = false;
        if (region == null)
        {
            regionValid = true;
        }
        else
        {
            regionValid = System.Convert.ToBoolean(this.Region.Encompasses(region));
        }

        if (dateValid)
        {
            if ((qty >= this.OptMin & qty <= this.OptMax) || qty == 0)
            {
                if (regionValid)
                {
                    foreach (var o in this.Options.Values)
                    {
                        if (o.ProdRef == prodref || prodref == "")
                        {
                            //avalanche gives a discount as a percentage of list price
                            returnValue.Add(o);
                        }
                    }
                }
            }
        }

        Pacc("getAvalancheOptions");

        return returnValue;
    }

    public void Update()
    {

        object sql = null;
        sql = "UPDATE Avalance SET ";
        sql += "opgRef=" + System.Convert.ToString(this.OPGref) + ",";
        //  sql$ &= "prodRef=" & Me.ProdRef & ","
        sql += "FK_region_id=" + this.Region.ID + ",";
        sql += "validFrom=" + da.universaldate(this.ValidFrom) + ",";
        sql += "validTo=" + da.universaldate(this.ValidTo) + ",";
        // sql$ &= "LPDiscountPercent =" & Me.LPDiscountPercent & ","
        sql += "optMin=" + System.Convert.ToString(this.OptMin) + ",";
        sql += "optMax=" + System.Convert.ToString(this.OptMax) + "";
        sql += " WHERE ID = " + System.Convert.ToString(this.ID);

        da.DBExecutesql(sql);

        //TODO : - need to update iq.i_OPGref

    }

    public ClsAvalancheOPG(string opgRef, clsRegion Region, DateTime validFrom, DateTime ValidTo, int optMin, int optMax, DataTable writecache = null)
    {

        this.OPGref = opgRef;
        //Me.ProdRef = prodRef
        this.Region = Region;
        this.ValidFrom = validFrom;
        this.ValidTo = ValidTo;
        this.OptMin = optMin;
        this.OptMax = optMax;
        this.Options = new Dictionary<int, ClsAvalancheOption>();

        if (writecache == null)
        {

            object sql = null;
            sql = "INSERT INTO avalancheOPG (opgref,optmin,optmax,validFrom,validTo,fk_region_id) ";
            sql += "VALUES (" + opgRef + "," + System.Convert.ToString(optMin) + "," + System.Convert.ToString(optMax) + "," + da.universaldate(validFrom) + "," + da.universaldate(ValidTo) + "," + Region.ID + ")";

            this.ID = System.Convert.ToInt32(da.DBExecutesql(sql, true));
            iq.AvalancheOPGs.Add(this.ID, this);

        }
        else
        {

            System.Data.DataRow row = default(System.Data.DataRow);
            row = writecache.NewRow();

            row["opgref"] = this.OPGref;
            //row("prodref") = Me.ProdRef
            row["optmin"] = this.OptMin;
            row["optmax"] = this.OptMax;
            //row("LPDiscountPercent") = Me.LPDiscountPercent
            row["validFrom"] = this.ValidFrom;
            row["validTo"] = this.ValidTo;
            row["fk_region_id"] = this.Region.ID;

            writecache.Rows.Add(row);

        }

        iq.i_OpgRef.Add(opgRef, this);

        //If a system has a descendant price with  offer which points to a pool contaiining the system

    }
    public ClsAvalancheOPG()
    {

        this.ID = -1;


    }


    public ClsAvalancheOPG(int ID, string opgRef, clsRegion Region, DateTime validFrom, DateTime ValidTo, int optMin, int optMax)
    {

        //the OPG's don't carry a set of systems - becuase the systems have the OPG's added (product.avalancheOPGs)... which is how they're needed
        //(we typically want to know the OPG's for a system - not the systems for an OPG)

        this.ID = ID;
        this.OPGref = opgRef;
        this.Region = Region;
        this.ValidFrom = validFrom;
        this.ValidTo = ValidTo;
        this.OptMin = optMin;
        this.OptMax = optMax;

        iq.AvalancheOPGs.Add(this.ID, this);
        iq.i_OpgRef.Add(opgRef, this);

        this.Options = new Dictionary<int, ClsAvalancheOption>();

    }


}